File: 567. Permutation in String.cpp
class Solution {
public:
	bool checkInclusion(string s1, string s2) {
		int n1 = s1.size(), n2 = s2.size();
		if (n1 > n2) return false;
        vector<int> v1(26, 0), v2(26, 0);
        for (int i = 0; i < n1; i++) {
            v1[s1[i]-'a']++;
            v2[s2[i]-'a']++;
        }

		for (int i = n1; i < n2; i++) {
            if (v1==v2) return true;
            v2[s2[i-n1] - 'a']--;
            v2[s2[i] - 'a']++;

		}
		return v1==v2;
	}
};

File: python/567. Permutation in String.py
import collections

class Solution(object):
    def checkInclusion(self, s1, s2):
        """
        :type s1: str
        :type s2: str
        :rtype: bool
        """
        k, n2 = len(s1), len(s2)
        if k > n2: return False

        pt = collections.defaultdict(int)
        for i in range(k):
            pt[s1[i]] += 1

        count = collections.defaultdict(int)
        for i in range(n2):
            count[s2[i]] += 1
            if i < k-1: continue
            # print i, count

            if pt==count: return True
            last = s2[i-k+1]
            count[last] -= 1
            if count[last] == 0:
                del count[last]
                
        return False

if __name__ == '__main__':
    s1 = "ab"
    s2 = "eidbaooo"
    su = Solution()
    print su.checkInclusion(s1, s2)